Mathlib/Tactic/Find.lean

```diff
@@ -39,96 +102,249 @@ private partial def matchHyps : List Expr → List Expr → List Expr → MetaM
+      terms := terms.push (true, t)
+    | `(find_pattern| $s:term) => do
+      let t ← Lean.Elab.Term.elabTerm s none
+      terms := terms.push (false, t)
+    | _ => throwErrorAt arg "unexpected argument to #find"
+  pure {idents, name_pats, terms}
+
+open Command
+
+/--
+The `#find` command finds definitions and lemmas by various ways, which can be combined:
+
+1. By constant:
+   ```lean
+   #find Real.sin tsum
+   ```
+   finds all lemmas that somehow mention the sine function and an infinite series sum.
+
+2. By lemma name substring:
+   ```lean
+   #find tsum "two"
+   ```
+   finds all lemmas about `tsum` that have `"two"` as part of the lemma _name_.
+
+3. By subexpression:
+   ```lean
+   #find tsum (_ * _ ^ _)
+   ```
+   finds all lemmas that mention `tsum` and where somewhere in the lemma statement there is a
+   product with a power as the second argument. The pattern can also be non-linear, as in
+   ```lean
+   #find (Real.sqrt ?a * Real.sqrt ?a)
+   ```
+
+4. By conclusion:
+   ```lean
+   #find ⊢ (tsum _ = _ * tsum _)
+   ```
+   finds all lemmas where the conclusion (the subexpression to the right of all `→` and `∀`) has the
+   given shape. If the pattern has hypotheses, then they are matched against the hypotheses of
+   the lemma in any order:
+   ```lean
+   #find ⊢ (_ < _ → tsum _ < tsum _)
+   ```
+   will find `tsum_lt_tsum` even though the hypotheses `f i < g i` is not the last.
+
+The command will show the list of results in the info view, where you can cover to see theirs types
+or click to go to their definition.
+
+If you pass more than one such search filter, it will list lemmas matching _all_ of them. At least
+some filter must mention a concrete name, because `#find` maintains an index of which lemmas
+mention which other constants. This is also why the _first_ use of `#find` will be somewhat slow
+(typically less than half a minute with all of `Mathlib` imported), but subsequent uses are faster.
+
+It may be useful to open a scratch file, `import Mathlib`, and use `#find` there, this way you will
+find lemmas that you have not yet imported, and the cache will stay up-to-date.
+
+Inside tactic proofs, the `#find` tactic can be used instead.
+-/
+elab s:"#find " args:find_pattern* : command => liftTermElabM do
+  profileitM Exception "find" (← getOptions) do
+    match ← find (← parseFindPatterns args) with
+    | .error warn =>
+      Lean.logWarningAt s warn
+    | .ok (summary, hits) =>
+      Lean.logInfo $ summary ++ (← MessageData.bulletListOfConsts hits)
+
+/--
 Tactic version of the `#find` command.
-See also the `find` tactic to search for theorems matching the current goal.
+See also the `apply?` tactic to search for theorems matching the current goal.
 -/
-elab "#find " t:term : tactic => do
-  let t ← Term.elabTerm t none
-  Term.synthesizeSyntheticMVars (mayPostpone := false) (ignoreStuckTC := true)
-  findType t
+elab s:"#find " args:find_pattern+ : tactic => do
```

Comment:
```suggestion
elab s:"find " args:find_pattern+ : tactic => do
```
Afaik `#` ought to be used only for commands...but maybe `#find` is more discoverable. Maybe we should allow both `find` and `#find` as tactic syntax in case people assume they have to strip `#` off when in tactic syntax. It shouldn't appear in committed code anyway, so imo whatever lets people find it most easily is best.

Reply:
I don't think I added the tactic. `#` is intended to mark temporary commands, I don't see why the marker wouldn't work for tactics as well. But it would be good to check for consistency with other std4/mathlib4 tactics.

Reply:
Is there a reason why commands can't simply always work in tactics? (Probably yes, but in curious)

Reply:
I don't think that makes sense for most commands, for example a `def` would not actually be scoped inside the current proof.

Reply:
Right, I guess I'm thinking of the #-class of commands. But if one has to decide for each command, then doing it like for #find suffices of course.

Reply:
So what is the conclusion here? Leave the `#find` “tactic” in (because it was there before this PR), or remove it? (I’m happy with either)

Reply:
I've started a thread about this [on zulip](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.236363.20.20feat.3A.20More.20to.20.23find!). :)